Front-end implementation and File upload functionality

emidailyd opened the pull request.

emidailyd committed:
Added primitive File Upload functionality, basic User Interface

emidailyd committed:
Vite + TypeScript

Changed JavaScript to Vite & TypeScript.

augisbud commented on client/src/App.tsx:
```diff
@@ -0,0 +1,55 @@
+import React, { useState, ChangeEvent, FormEvent } from 'react';
```
'React' is declared but its value is never read.

augisbud commented on client/README.md:
```diff
@@ -0,0 +1,50 @@
+# React + TypeScript + Vite
```
README.md from client directory can be removed.

augisbud commented on client/public/vite.svg:
Can be removed

augisbud commented on client/src/App.css:
Too much leftover from example.

After SASS configuration, should be renamed to `App.module.scss`

and the contents can be changed to:
```css
.wrapper {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
}

.title {
  font-size: 3.2em;
  line-height: 1.1;
}

.button {
  border-radius: 8px;
  border: 1px solid transparent;
  padding: 0.6em 1.2em;
  font-size: 1em;
  font-weight: 500;
  font-family: inherit;
  background-color: #1a1a1a;
  cursor: pointer;
  transition: border-color 0.25s;
}
.button:hover {
  border-color: #646cff;
}
.button:focus, .button:focus-visible {
  outline: 4px auto -webkit-focus-ring-color;
}

@media (prefers-color-scheme: light) {
  :root {
    color: #213547;
    background-color: #ffffff;
  }
  .button {
    background-color: #f9f9f9;
  }
}
```

instead of extending specific HTML type, we would rather create classes for individual button and etc, that can be later reused at our own discretion.

augisbud commented on client/src/App.tsx:
```diff
@@ -0,0 +1,55 @@
+import React, { useState, ChangeEvent, FormEvent } from 'react';
+
```
After moving to usage of classes we should:
```js
import styles from './App.module.scss';
```

augisbud commented on client/src/App.tsx:
```diff
@@ -0,0 +1,55 @@
+    };
+
+    return (
+        <div>
```
Subsequently, the TSX element needs to have the classes specified from styles, which can be achieved in the following way:

```js
return (
        <div className={styles.wrapper}>
            <h1 className={styles.title}>Upload a File</h1>
            <form onSubmit={handleSubmit}>
                <input type="file" onChange={handleFileChange} />
                <button className={styles.button} type="submit">Upload</button>
            </form>
        </div>
    );
```

augisbud commented on client/src/index.css:
```diff
@@ -0,0 +1,68 @@
+:root {
```
We should rename it to `globals.scss` and remove no longer used properties, which would make our file like so:
```css
@use '../src/style/variables.scss' as *;
@use '../src/style/typography.module.scss' as *;

html,
body {
    font-family: Inter, system-ui, Avenir, Helvetica, Arial, sans-serif;
    line-height: 1.5;
    font-weight: 400;

    color-scheme: light dark;
    color: rgba(255, 255, 255, 0.87);
    background-color: #242424;

    font-synthesis: none;
    text-rendering: optimizeLegibility;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
}

h1 {
    @extend %h1;
}

h2 {
    @extend %h2;
}

h3 {
    @extend %h3;
}

h4 {
    @extend %h4;
}

h5 {
    @extend %h5;
}

p {
    @extend %p;
}
```

augisbud commented on client/src/App.tsx:
```diff
@@ -0,0 +1,55 @@
+import React, { useState, ChangeEvent, FormEvent } from 'react';
+
+function App() {
```
According to modern javascript, it should be `export const App = () => {`

augisbud commented on client/src/main.tsx:
```diff
@@ -0,0 +1,10 @@
+import { StrictMode } from 'react'
+import { createRoot } from 'react-dom/client'
+import App from './App.tsx'
```
After updating the export in App.tsx, this will have to be updated to `import { App } from './App.tsx'`

augisbud commented on QuizCraft.Domain.API/Program.cs:
```diff
@@ -1,12 +1,30 @@
+    options.AddPolicy("AllowAllOrigins",
+        builder =>
+        {
+            builder.AllowAnyOrigin()
```
Maybe lets allow only `localhost:3000` for now, later we can change to a deployed URL.

augisbud commented on client/.gitignore:
Maybe, lets have a single .gitignore for both the .NET project & the TS project?

augisbud reviewed (commented):
I would say that we should rename the client folder to something like QuizCraft.Frontend

We would probably want to use SCSS for the additional functionality

`npm i sass`, then

vite.config.ts should be updated to:
```js
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [react()],
  server: {
    port: 3000
  },
  css: {
    modules: {
      localsConvention: "camelCase"
    }
  }
})
```

augisbud commented:
Also we can perform this separately, but it would be logical to add a GitHub Actions Template to Build & Publish UI image or are we considering deploying it on a S3 bucket in AWS using cloudfront? Would need to analyse the costs of such approach, leaning more towards using NGINX in a container to serve the static files.

emidailyd committed:
Implemented suggestions

Deleted unnecessary files and code.
Added SCSS.
Merged frontend's .gitignore into the main one.
Changed CORS policy.

emidailyd committed:
Added file processing logic

emidailyd commented:
The highlighted problems have been resolved (hopefully). I have also added the file processing logic. Currently the Unit and Integration tests have not yet been updated with new functionality that has been added. As for the deployment strategy, perhaps we can discuss it later on in the development. 
<img src="https://media1.giphy.com/media/c4u2gld3Or69i/giphy.gif"/>
